src/coins.h

```diff
@@ -103,8 +106,29 @@ class Coin
  */
 struct CCoinsCacheEntry
 {
+private:
+    /**
+     * These are used to create a doubly linked list of flagged entries.
+     * They are set in AddFlags and unset in ClearFlags.
+     * A flagged entry is any entry that is either DIRTY, FRESH, or both.
+     *
+     * DIRTY entries are tracked so that only modified entries can be passed to
+     * the parent cache for batch writing. This is a performance optimization
+     * compared to giving all entries in the cache to the parent and having the
+     * parent scan for only modified entries.
+     *
+     * FRESH-but-not-DIRTY coins can not occur in practice, since that would
+     * mean a spent coin exists in the parent CCoinsView and not in the child
+     * CCoinsViewCache. Nevertheless, if a spent coin is retrieved from the
+     * parent cache, the FRESH-but-not-DIRTY coin will be tracked by the linked
+     * list and deleted when Sync or Flush is called on the CCoinsViewCache.
+     */
+    CoinsCachePair* m_prev{nullptr};
```

Comment:
Not sure if it helps, just an observation: `m_prev` only uses `.second`, so it might as well be a `CCoinsCacheEntry*`
```patch
Index: src/coins.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/coins.h b/src/coins.h
--- a/src/coins.h	(revision c36363f6b24c7ab2afe198d9855f507ddf096e1f)
+++ b/src/coins.h	(date 1719730674870)
@@ -123,7 +123,7 @@
      * parent cache, the FRESH-but-not-DIRTY coin will be tracked by the linked
      * list and deleted when Sync or Flush is called on the CCoinsViewCache.
      */
-    CoinsCachePair* m_prev{nullptr};
+    CCoinsCacheEntry* m_prev{nullptr};
     CoinsCachePair* m_next{nullptr};
     uint8_t m_flags{0};
 
@@ -157,9 +157,9 @@
     inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& head)
     {
         if (!m_flags && flags) {
-            m_prev = &head;
+            m_prev = &head.second;
             m_next = head.second.m_next;
-            if (m_next) m_next->second.m_prev = &self;
+            if (m_next) m_next->second.m_prev = &self.second;
             head.second.m_next = &self;
         }
         m_flags |= flags;
@@ -168,8 +168,9 @@
     {
         if (m_flags) {
             if (m_next) m_next->second.m_prev = m_prev;
-            m_prev->second.m_next = m_next;
-            m_prev = m_next = nullptr;
+            m_prev->m_next = m_next;
+            m_prev = nullptr;
+            m_next = nullptr;
         }
         m_flags = 0;
     }
```

Reply:
Done, nice catch!